### PR TITLE
Move get_array_module to backends from cuda

### PR DESCRIPTION
**Pull request by hvy**
Moved `backends.cuda.get_array_module` to `backends.get_array_module`. 

As `get_array_module` may return any "module", it's more appropriate to place it at a higher level than under a a particular module.

### CODE REVIEW

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
If the new namespace looks fine, I'll replace the calls.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Will add tests.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
Introduced a chainer/backend.py as discussed offline. The motivation is that we do not want to introduce too much logic into a module of plural name. A module of plural name should, if possible only contain instances of the plural description.

**Comment by hvy** [comment]
jenkins, test this please.

File: chainer/backends/cuda.py [hunk 1]
```diff
@@ -533,6 +533,11 @@ def get_array_module(*args):
     it will return their data arrays' array module for
     :class:`~chainer.Variable` arguments.
 
+    .. note::
```
Comment (kmaehashi) [inline chainer/backends/cuda.py]: Please use `.. deprecated:: v5.0.0` instead.

File: chainer/backends/cuda.py [hunk 2]
```diff
@@ -541,12 +546,9 @@ def get_array_module(*args):
         the arguments.
 
     """
-    if available:
-        args = [arg.data if isinstance(arg, chainer.variable.Variable) else arg
-                for arg in args]
-        return cupy.get_array_module(*args)
-    else:
-        return numpy
+    warnings.warn('backends.cuda.get_array_module is deprecated. Please use '
```
Comment (kmaehashi) [inline chainer/backends/cuda.py]: I think it's too early to emit warning, considering that this API is used in many places. How about doing it in v6?

File: chainer/backends/cuda.py (hunk 1 above)
Comment (hvy) [reply chainer/backends/cuda.py]: Fixed.

File: chainer/backends/cuda.py (hunk 2 above)
Comment (hvy) [reply chainer/backends/cuda.py]: Thanks for the suggestion. I also think that's okay so removed the warning.

**Comment by hvy** [comment]
Thanks you for the review, PTAL.

File: tests/chainer_tests/backends_test/test_init.py [hunk 3]
```diff
@@ -80,4 +82,30 @@ def test_fail_on_invalid_dst(self):
             backends.copyto(dst, src)
 
 
+class TestGetArrayModule(unittest.TestCase):
```
Comment (hvy) [inline tests/chainer_tests/backends_test/test_init.py]: Noticed I should move these test to a new test file for backend.py.

Comment (hvy) [reply tests/chainer_tests/backends_test/test_init.py]: Fixed.

**Comment by hvy** [comment]
jenkins, test this please.

File: tests/chainer_tests/test_backend.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    def test_get_array_module_for_cupy_array(self):
+        xp = backend.get_array_module(cuda.cupy.array([]))
+        assert xp is cuda.cupy
+        assert xp is not numpy
+
+    @attr.gpu
+    def test_get_array_module_for_cupy_variable(self):
+        xp = backend.get_array_module(chainer.Variable(cuda.cupy.array([])))
+        assert xp is cuda.cupy
+        assert xp is not numpy
```
[the remark this answers is not in the document]
Comment (hvy) [reply tests/chainer_tests/test_backend.py]: Thanks, added the missing line.

**Comment by hvy** [comment]
Resolved conflicts.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
Resolved conflicts. PTAL.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by niboshi** [comment]
LGTM

@kmaehashi I merge this PR as seemingly all of the previous review comments have been resolved.

**Pull request merged**